usrp_drivers/n200/n200_driver.cpp

```diff
@@ -382,245 +398,253 @@ std::string random_string( size_t length )
-  computation_socket.connect("ipc:///tmp/feeds/1");
+  zmq::socket_t rx_dsp_socket(driver_c, ZMQ_PAIR);
+  ERR_CHK_ZMQ(rx_dsp_socket.connect("ipc:///tmp/feeds/1"))
 
 
 
-  //Allows for set up of SUB/PUB socket
+  //Sleep to handle "slow joiner" problem
+  //http://zguide.zeromq.org/php:all#Getting-the-Message-Out
   sleep(1);
 
   while (1) {
-    std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
     radarctrl_socket.recv(&request);
-    std::chrono::steady_clock::time_point end = std::chrono::steady_clock::now();
 
-    begin = std::chrono::steady_clock::now();
-    driverpacket::DriverPacket dp;
+    auto begin_deserialize = std::chrono::steady_clock::now();
+    driverpacket::DriverPacket driver_packet;
     std::string msg_str(static_cast<char*>(request.data()), request.size());
-    dp.ParseFromString(msg_str);
+    driver_packet.ParseFromString(msg_str);
 
-    std::cout << "Control " << dp.sob() << " " << dp.eob()
-      << " " << dp.channels_size() << std::endl;
+    DEBUG_MSG("Control " << driver_packet.sob() << " " << driver_packet.eob()
+      << " " << driver_packet.channels_size());
 
-    if ((dp.sob() == true) && (dp.eob() == true)) {
+    //TODO(keith): thinking about moving this err chking to transmit
+    if (driver_packet.samples_size() != driver_packet.channels_size()) {
       // TODO(keith): throw error
     }
 
-    if (dp.samples_size() != dp.channels_size()) {
-      // TODO(keith): throw error
-    }
-
-    for (int i = 0; i < dp.samples_size(); i++) {
-      if (dp.samples(i).real_size() != dp.samples(i).imag_size()) {
+    for (int i = 0; i < driver_packet.samples_size(); i++) {
+      if (driver_packet.samples(i).real_size() != driver_packet.samples(i).imag_size()) {
         // TODO(keith): throw error
       }
-    }
+    }// REVIEW #6 TODO: Check that seq num is not 0 (assuming we change it to start at 1 -either way check that there is a seq num) TODO: Check that timetosendsamples is set - maybe check if in future? TODO: Check that channels, samples, ctrfreq, rxrate, txrate are set if sob==true
+     // REPLY Pretty much all of this is getting checked elsewhere. How to check timetosendsamples?
+    driver_packet_pub_socket.send(request);
 
-    packet_socket.send(request);
+    auto end_deserialize = std::chrono::steady_clock::now();
 
-    end = std::chrono::steady_clock::now();
-    std::cout << "Time difference to deserialize and send in control = "
-      << std::chrono::duration_cast<std::chrono::microseconds>
-                                                  (end - begin).count()
-      <<std::endl;
+    auto time_diff = std::chrono::duration_cast<std::chrono::microseconds>
+                                                  (end_deserialize - begin_deserialize).count();
+    DEBUG_MSG("Time difference to deserialize and send in control = " << time_diff << "us");
 
-    if (dp.eob() == true) {
-      zmq::message_t ack, data, size;
+    if (driver_packet.eob() == true) {
```

Comment:
Inside this if statement we could do:
1) send shr_mem address to rx_dsp so it can start doing it's thing right away
2)  get ack data from receive thread/transmit thread 
3) deserialize ack data from rx/tx threads 
4) check for errors in the deserialized ack data
5) Do something about errors 

Alternatively if it is quick to deserialize then do that first, before sending shr_mem address to rx_dsp,because an error in tx is going to affect what we want rx_dsp to do as well

TODO: discuss this as a group